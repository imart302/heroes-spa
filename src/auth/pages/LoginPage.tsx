import React, { useContext } from 'react';
import { useNavigate } from 'react-router-dom';
import { AuthContext } from '../context';

export const LoginPage = () => {

  const authContext = useContext(AuthContext);
  const navigate = useNavigate();

  const onLogin = () => {
    authContext.login?.({
      id: '1',
      user: 'Ivan',
    })

    const lastPath = localStorage.getItem('lastPath') || '/';

    navigate(lastPath, {
      replace: true
    });
  }

  return (
    <div className="contianer mt-5">
      <h1>Login</h1>
      <hr />

      <button className="btn btn-primary" onClick={onLogin}>
        Login
      </button>
    </div>
  );
};
