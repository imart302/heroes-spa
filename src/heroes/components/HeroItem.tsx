import React from 'react';
import { IHero } from '../models/hero';

export const HeroItem = ({ hero }: { hero: IHero }) => {
  return (

    <li>
      <h6>{hero.superhero}</h6>
    </li>

  );
};
