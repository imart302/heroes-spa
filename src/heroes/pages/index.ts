export * from './DCPage';
export * from './Hero';
export * from './MarvelPage';
export * from './Search';
export * from './index';
