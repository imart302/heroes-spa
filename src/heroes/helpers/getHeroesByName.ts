import { heroes } from "../data/heroes"


export const getHeroesByName = (name: string) => {
  
  const cName = name.toLowerCase().trim();

  return heroes.filter( hero => {
    return hero.superhero.toLowerCase().includes(cName);
  });
}