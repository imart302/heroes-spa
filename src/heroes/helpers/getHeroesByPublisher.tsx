import { heroes } from "../data/heroes";


export const getHeroesByPublisher = (publisher: 'Marvel Comics' | 'DC Comics') => {

  return heroes.filter(hero => hero.publisher == publisher);

}