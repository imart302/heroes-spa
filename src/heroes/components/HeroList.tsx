import React, { useMemo } from 'react';
import { getHeroesByPublisher } from '../helpers/getHeroesByPublisher';
import { HeroCard } from './HeroCard';
import { HeroItem } from './HeroItem';

export const HeroList = ({
  publisher,
}: {
  publisher: 'Marvel Comics' | 'DC Comics';
}) => {
  
  const heroes = useMemo(() => {
    return getHeroesByPublisher(publisher);
  }, [publisher]);

  return (
    <div className="row rows-cols-1 row-cols-md-3 g-3">
      {heroes.map((hero) => (
        <HeroCard key={hero.id} hero={hero} />
      ))}
    </div>
  );
};
