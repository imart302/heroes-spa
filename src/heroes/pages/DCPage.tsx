import React from 'react'
import { HeroList } from '../components/HeroList'

export const DCPage = () => {
  return (
    <>
      <HeroList publisher='DC Comics' />
    </>
  )
}
