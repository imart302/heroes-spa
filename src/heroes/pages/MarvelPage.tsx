import React, { useContext } from 'react'
import { AuthContext } from '../../auth/context'
import { HeroList } from '../components/HeroList'

export const MarvelPage = () => {

  return (
    <>
      <HeroList publisher='Marvel Comics' />
    </>
  )
}
