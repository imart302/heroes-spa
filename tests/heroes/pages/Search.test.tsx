import { fireEvent, render, screen } from '@testing-library/react';
import { MemoryRouter, useNavigate } from 'react-router-dom';
import { vi } from 'vitest';
import React from 'react';
import { Search } from '../../../src/heroes/pages';
import { HeroCard } from '../../../src/heroes/components/HeroCard';
import { IHero } from '../../../src/heroes/models/hero';
import { getHeroesByName } from '../../../src/heroes/helpers/getHeroesByName';

vi.mock('react-router-dom', async () => {
  const mod = await vi.importActual<typeof import('react-router-dom')>(
    'react-router-dom'
  );
  return {
    ...mod,
    useNavigate: vi.fn(),
  };
});

vi.mock('../../../src/heroes/helpers/getHeroesByName', () => {
  return {
    getHeroesByName: vi.fn(),
  };
});

vi.mock('../../../src/heroes/components/HeroCard', () => {
  return {
    HeroCard: vi.fn(),
  };
});

describe('Test on <Search />', () => {
  beforeEach(() => {
    vi.resetAllMocks();
  });

  test('Should show correct with default values', () => {
    const { container } = render(
      <MemoryRouter>
        <Search />
      </MemoryRouter>
    );

    expect(container).toMatchSnapshot();
  });

  test('Should show Batman with the query', () => {
    vi.mocked(HeroCard).mockImplementation(({ hero }: { hero: IHero }) => {
      return <h1 data-testid="hero-test-id">{hero.superhero}</h1>;
    });

    const heroes = [
      {
        alter_ego: 'batman',
        characters: 'bruce',
        first_appearance: '1234',
        id: 'dc-batman',
        publisher: 'DC Comics',
        superhero: 'Batman Test',
      },
    ];

    vi.mocked(getHeroesByName).mockReturnValue(heroes);

    render(
      <MemoryRouter initialEntries={['/search?q=batman']}>
        <Search />
      </MemoryRouter>
    );

    const batmanEl = screen.getByText(heroes[0].superhero);
    expect(batmanEl).toBeDefined();
  });

  test('Should not display (none) info and alert not found hero when search', () => {
    vi.mocked(HeroCard).mockImplementation(({ hero }: { hero: IHero }) => {
      return <h1 data-testid="hero-test-id">{hero.superhero}</h1>;
    });

    const heroes = [
      {
        alter_ego: 'batman',
        characters: 'bruce',
        first_appearance: '1234',
        id: 'dc-batman',
        publisher: 'DC Comics',
        superhero: 'Batman Test',
      },
    ];

    vi.mocked(getHeroesByName).mockReturnValue(heroes);

    render(
      <MemoryRouter initialEntries={['/search?q=batman']}>
        <Search />
      </MemoryRouter>
    );

    const infoSearch = screen.getByLabelText('info-search-hero');
    const alertNotFound = screen.getByLabelText('alert-hero-notfound');

    expect(infoSearch.style.display).toBe('none');
    expect(alertNotFound.style.display).toBe('none');
  });

  test('Should display alert when no hero found', () => {
    vi.mocked(HeroCard).mockImplementation(({ hero }: { hero: IHero }) => {
      return <h1 data-testid="hero-test-id">{hero.superhero}</h1>;
    });

    const heroes = [];

    vi.mocked(getHeroesByName).mockReturnValue(heroes);

    render(
      <MemoryRouter initialEntries={['/search?q=batman']}>
        <Search />
      </MemoryRouter>
    );

    const alertNotFound = screen.getByLabelText('alert-hero-notfound');

    expect(alertNotFound.style.display).toBe(''); 
  });

  test('Should navigate with the query when submit form', () => {
    const navigate = vi.fn();
    vi.mocked(useNavigate).mockReturnValue(navigate);

    render(
      <MemoryRouter>
        <Search />
      </MemoryRouter>
    );

    const searchText = 'Batman';
    const input = screen.getByRole<HTMLInputElement>('textbox');
    fireEvent.change(input, { target: { value: searchText } });
    const form = screen.getByRole<HTMLFormElement>('form');

    fireEvent.submit(form);

    expect(navigate).toBeCalledWith(`?q=${searchText}`);
  });

});
