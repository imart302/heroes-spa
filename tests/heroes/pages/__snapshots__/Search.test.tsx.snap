// Vitest Snapshot v1

exports[`Test on <Search /> > Should show correct with default values 1`] = `
<div>
  <h1>
    Search
  </h1>
  <hr />
  <div
    class="row"
  >
    <div
      class="col-5"
    >
      <h4>
        Seaching
      </h4>
      <hr />
      <form
        aria-label="form"
      >
        <input
          autocomplete="off"
          class="form-control"
          name="searchText"
          placeholder="Search a hero"
          type="text"
        />
        <button
          class="btn btn-outline-primary mt-1"
          type="submit"
        >
          Search
        </button>
      </form>
    </div>
    <div
      class="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        aria-label="info-search-hero"
        class="alert alert-primary"
      >
        Search a hero
      </div>
      <div
        aria-label="alert-hero-notfound"
        class="alert alert-danger"
        style="display: none;"
      >
        No hero found
      </div>
    </div>
  </div>
</div>
`;
